pkg/autodiscovery/common/utils/prometheus_kubelet.go

```diff
@@ -28,18 +29,41 @@ func ConfigsForPod(pc *types.PrometheusCheck, pod *kubelet.Pod) []integration.Co
 
 	instances, found := buildInstances(pc, pod.Metadata.Annotations, namespacedName)
 	if found {
-		for _, container := range pod.Status.GetAllContainers() {
-			if !pc.AD.MatchContainer(container.Name) {
-				log.Debugf("Container '%s' doesn't match the AD configuration 'kubernetes_container_names', ignoring it", container.Name)
+		for _, containerStatus := range pod.Status.GetAllContainers() {
+			if !pc.AD.MatchContainer(containerStatus.Name) {
+				log.Debugf("Container '%s' doesn't match the AD configuration 'kubernetes_container_names', ignoring it", containerStatus.Name)
 				continue
 			}
+			// If `prometheus.io/port` annotation has been provided, let’s keep only the container that declares this port.
+			if portFromAnnotationString, found := pod.Metadata.Annotations[types.PrometheusPortAnnotation]; found {
+				portFromAnnotation, _ := strconv.Atoi(portFromAnnotationString)
+				var containerSpec *kubelet.ContainerSpec
+				for _, ctr := range pod.Spec.Containers {
```

Comment:
In most cases, the number of containers in a pod should remain low enough that this isn’t an issue.
But it’s definitively better to protect against pathological cases.
I’ve rewritten the algorithm to use maps and to avoid the two nested loops: https://github.com/DataDog/datadog-agent/pull/16006/commits/bd6378f49af70f10ae21846f989bc834294e44e5